testsuite/forge/src/backend/local/mod.rs

```diff
@@ -1,85 +1,124 @@
 // Copyright (c) The Diem Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-use crate::{Factory, Result, Swarm};
-use anyhow::{bail, Context};
-use std::{env, path::PathBuf, process::Command};
+use crate::{Factory, Result, Swarm, Version};
+use std::{
+    collections::HashMap,
+    num::NonZeroUsize,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
 
+mod cargo;
 mod node;
 mod swarm;
 pub use node::LocalNode;
-pub use swarm::{LocalSwarm, LocalSwarmBuilder};
+pub use swarm::{LocalSwarm, LocalSwarmBuilder, SwarmDirectory};
+
+#[derive(Clone, Debug)]
+pub struct LocalVersion {
+    revision: String,
+    bin: PathBuf,
+    version: Version,
+}
+
+impl LocalVersion {
+    fn bin(&self) -> &Path {
+        &self.bin
+    }
+
+    fn version(&self) -> Version {
+        self.version.clone()
+    }
+}
 
 pub struct LocalFactory {
-    diem_node_bin: String,
+    versions: Arc<HashMap<Version, LocalVersion>>,
 }
 
 impl LocalFactory {
```

Comment:
just a point wanna through out to discuss:
I think I can understand the different constructors here for LocalFactory, 1. the latest version of current workspace. 2. specified version based on revision. 3. both of above cases for compatibility test.

First point is, from my personal perspective, I felt a little bit confusion about the naming at the first time before I dig deep into the code.
Second point is: From above we can see the only case we need two versions is for comp test, do we really need an Arc hashmap to save versions. What if we do a unified fn by pass in a revision include HEAD, to return only a version each time? then we can call `with_upstream_and_workspace` only for comp test case only?
for `LocalFactory` itself we may can have an Optional field `another_version`(ignore the naming) and default to None?

Generally it looks awesome 👍 just in case to discuss the code route open mind here :)

Reply:
I thought a great deal about all of this when setting up this construct. The main benefit with how this PR implement it is that if you're running a bunch of non-compat tests then its still just as easy to instantiate a LocalFactory as it was before and if you're wanting to do some compat tests you have the flexibility to to test with more than just 2 versions (which once i told @sausagee was possible was very excited by the prospect).

Naming-wise, yeah maybe the constructor names aren't the best. Its easy enough to improve that

Reply:
gotcha, if we prospect to support multi versions mixture comp test this is a good move. 👍